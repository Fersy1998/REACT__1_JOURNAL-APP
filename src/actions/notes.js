import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote=()=>{
    return async(dispatch, getState)=>{
        const uid=getState().auth.uid;
        const newNote={
            title:'',
            body:'',
            date:new Date().getTime(),
            url:''
        }
       
        try {
            const doc= await db.collection(`${uid}/journal/notes`).add(newNote);
            dispatch(activeNote(doc.id, newNote));
        } catch (error) {
            return error;
        }
        
        //dispatch(AddNewNote(doc.id, newNote));
        //dispatch(setNotes(getState().notes.notes));
    }
}

export const activeNote=(id, note)=>({
    type: types.notesActive,
    payload:{
        id,
        ...note
    }
})
export const startLoadingNotes=(uid)=>{
    return async (dispatch)=>{
        const notes=await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}
export const setNotes=(notes)=>({
    type:types.notesLoad,
    payload:notes
})

export const startSaveNotes=(note)=>{
    return async(dispatch, getState)=>{
        const uid=getState().auth.uid;
        let noteToFirestore={...note};
        if(!note.url){
            delete note.url;
        }
        delete noteToFirestore.id;
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
        dispatch(deleteNote(note.id));
        dispatch(AddNewNote(note.id, {...note} ));
        dispatch(activeNote(note.id, {...note} ));
        Swal.fire('Saved', note.title, 'success' );
    }
}
export const AddNewNote=(id, note)=>({
    type:types.notesAddNew,
    payload:{
        id, ...note
    }
    
    })

//Recarga la nota editada solamente
export const refreshNote=(id, note)=>({
    type:types.notesUpdated,
    payload:{
        id,
        note:{
            ...note
        }
    }
})

export const startUploading=(file)=>{
    return async (dispatch, getState)=>{
        const {active: activenote}=getState().notes;
        Swal.fire({
            title:'Uploading',
            html:'<p>Please wait...</p>',
            allowOutsideClick:false,
            
        })
        const fileURL= await fileUpload(file);
        //console.log(fileURL);
        activeNote.url=fileURL;
        dispatch(startSaveNotes({...activenote, url:fileURL}));
        dispatch(refreshNote(activeNote.id, {...activeNote, url:fileURL} ));
        
        Swal.close();
    }
}
export const startDeleting=(id)=>{
    return async (dispatch, getState)=>{
        const {uid}=getState().auth;
        await db.doc(`${uid}/journal/notes/${id}`).delete();
        dispatch(deleteNote(id));
    }
}

export const deleteNote=(id)=>({
    type:types.notesDelete,
    payload: id
})
export const noteLogout=()=>({
    type:types.notesLogOutCleaning,
})