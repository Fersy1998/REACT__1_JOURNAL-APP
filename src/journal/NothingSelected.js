import React from 'react'

export const NothingSelected = () => {
  return (
    <div className='nothing__main-content '>
        <p className='animate__animated animate__headShake'>
            Select something<br />
            Create a new Entry!
        </p>
        <i className='fa fa-star fa-4x mt-5'></i>
    </div>
  )
}
