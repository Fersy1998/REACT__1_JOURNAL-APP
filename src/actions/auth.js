import { types } from "../types/types"
import Swal from 'sweetalert2'
import React from 'react';

import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";
import { Navigate, useNavigate } from "react-router-dom";
import { noteLogout } from "./notes";


export const startLoginEmailPassword=(email, password)=>{
    
    return (dispatch)=>{
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password)
        .then(async ({user})=>{
            //console.log(res);
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        }).catch((e)=>{
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error')
        })
       
    }
/*
    return (dispatch)=>{
        setTimeout(()=>{
            dispatch(login(email, password));
        
        }, 3500)
    }*/
}

export const startRegisterWithEmalPasswordName=(email, password, name)=>{
    return (dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then(async ({user})=>{   
            //console.log(user);
            await user.updateProfile({displayName:name});
            dispatch(login(user.uid, user.displayName));
        
    }).catch((e)=>{
        //dispatch(finishLoading());
        Swal.fire('Error', e.message, 'error')
    })
    }
}
export const startGoogleLogin=()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider.setCustomParameters({prompt: "select_account"}))
        /*.then(userCred=>{
            console.log(userCred);
        })*/
      .then(({user})=>{   
            console.log(user);
            dispatch(login(user.uid, user.displayName));
        }
        )
    }
}

export const login=(uid, displayname)=>{
    
    
    return {
        type:types.login,
        payload:{
            uid,
            displayname
        }
    }
}

export const startLogOut=()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut()
        dispatch(logOut());
        dispatch(noteLogout());
    }

}
export const logOut=()=>{
    return {
        type:types.logout
    } 
}
