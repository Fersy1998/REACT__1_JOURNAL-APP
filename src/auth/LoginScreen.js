import React, { useEffect, useState } from 'react'
import { Link, Navigate, useNavigate } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../hooks/useForm'
import {setError, removeError} from '../actions/ui';
import { startGoogleLogin, startLoginEmailPassword } from '../actions/auth'
import validator from 'validator';
export const LoginScreen = () => {
    
    const navigate=useNavigate();
    const dispatch=useDispatch()
    const {msgError, loading}=useSelector(state=>state.ui);
    const user=useSelector(state=>state.auth)
    
    const [formValues, handleInputChange]=useForm({
        email:'fersy@gmail.com',
        password:'1hwdhid'
    })
    const [userActive, setActiveUser] = useState(user);
    useEffect(() => {
        if(user?.uid){
            setActiveUser(true);
            navigate('/auth/journal', {
                replace:true
              }) ;
          }else{
            setActiveUser(false);
          }
    }, [setActiveUser])
    

    
    
    const {email, password}=formValues;
    const handleLogin=(e)=>{
        e.preventDefault();
        //dispatch(login(892183127, 'Fersy'));
        if (isFormValid()){
            dispatch(startLoginEmailPassword(email, password));
            navigate('/', {
                replace:true
              }) ;
        }
    }
    const handleGoogleLogin=()=>{
        if(isFormValid()){
            dispatch(startGoogleLogin());
            
        }
        
    }
    const isFormValid=()=>{
        if(!validator.isEmail(email)){
          dispatch(setError('Invalid email'));
          return false;
        }else if(password.length<6 ){
          dispatch(setError('password should have at least 6 characters and match'));
          return false
        }
        dispatch(removeError());
        return true
      
      }
  return (

    <div className='animate__animated animate__headShake'>
        <h3 className='auth__title'>
            Login
        </h3>
        <form onSubmit={handleLogin}>
            {
            msgError &&
            (<div className='auth__alert-error'>{msgError}</div>)
            }
            <input type='text' placeholder='email' onChange= {handleInputChange} name='email' className='auth__input' value={email}/>
            <input type='password' placeholder='password' onChange= {handleInputChange} name='password' className='auth__input' value={password}/>
            <button type='submit' className='btn btn-primary btn-block' disabled={loading}>Login</button><hr />
            <div className='auth__social-network'>
                <p>Login with social Network</p>
                <div className="google-btn" onClick={handleGoogleLogin}>
                    <div className="google-icon-wrapper">
                        <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                    </div>
                    <p className="btn-text">
                        <b>Sign in with google</b>
                    </p>
                </div>
            </div>
            <Link to='/auth/register' className='link'>
                Create New Acount
            </Link>
        </form>
        
    </div>

  )
}
