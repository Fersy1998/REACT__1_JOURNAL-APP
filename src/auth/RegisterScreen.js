import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../hooks/useForm'
import validator from 'validator';
import {setError, removeError} from '../actions/ui';
import { startRegisterWithEmalPasswordName } from '../actions/auth';
export const RegisterScreen = () => {

  const [formValues, handleInputChange]=useForm({name:'Fersy', email:'fersy@yahoo.com', password:'123', password2:'123'});
  const dispatch=useDispatch();
  
  //const state=useSelector(state=>state.ui); extrae el estado de Redux del navegador
  const {msgError}=useSelector(state=>state.ui);
  const {name, email, password, password2}=formValues;
  
  
  const handleRegister=(e)=>{
    e.preventDefault();
    
    if(isFormValid()){
      dispatch(startRegisterWithEmalPasswordName(email, password, name));
      //console.log(formValues);
    }
  
  }
  const isFormValid=()=>{
    if(name.trim().length===0){
      dispatch(setError('name is required'));
      return false;
    }else if(!validator.isEmail(email)){
      dispatch(setError('Invalid email'));
      return false;
    }else if(password!==password2 || password.length<6 ){
      dispatch(setError('password should have at least 6 characters and match'));
      return false
    }
    dispatch(removeError());
    return true
  
  }
  return (
    
    <div className='animate__animated animate__headShake'>
        <h3 className='auth__title' >
            Register
        </h3>
        <form onSubmit={handleRegister}>
            {
            msgError &&
            (<div className='auth__alert-error'>{msgError}</div>)
            }
            <input type='text' placeholder='name' name='name' className='auth__input'value={name} onChange={handleInputChange}/>
            <input type='text' placeholder='email' name='email' className='auth__input'value={email} onChange={handleInputChange}/>
            <input type='password' placeholder='password' name='password' className='auth__input' value={password} onChange={handleInputChange}/>
            <input type='password' placeholder='confirm password' name='password2' className='auth__input' value={password2} onChange={handleInputChange}/>
            <button type='submit' className='btn btn-primary btn-block'>Login</button><hr />
            <Link to='auth/login' className='link'>
                Already register?
            </Link>
        </form>
        
    </div>
 
  )
}
