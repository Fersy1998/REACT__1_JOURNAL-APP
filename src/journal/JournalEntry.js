import React from 'react'
import moment from 'moment'
import { useDispatch } from 'react-redux';
import { activeNote } from '../actions/notes';
export const JournalEntry = (note) => {
    const {id, title, date, body, url} =note;
    const noteDate=moment(date);
    const dispatch=useDispatch();
    const handleEntryClick=()=>{
        dispatch(activeNote(id, note));
    }
  return (
    <div className='journal__entry pointer' onClick={handleEntryClick}>
    
        { url &&
        
        (<div className='journal__entry-picture'
            style={{backgroundSize:'cover',
                backgroundImage:`url(${url})`
            }}
        ></div>)}
        <div className='journal__entry-body'>
            <p className='journal__entry-title'>{title}</p>
            <p className='journal__entry-content'>{body}</p>
        </div>
        <div className='journal__entry-date-box'>
            <span>{noteDate.format('dddd')}</span>
            <p>{noteDate.format('Do')}</p>
        
        </div>
    </div>
  )
}
