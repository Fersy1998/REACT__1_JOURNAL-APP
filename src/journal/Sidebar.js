import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useNavigate } from 'react-router-dom';
import { startLogOut } from '../actions/auth';
import { noteLogout, startNewNote } from '../actions/notes';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {
  const navigate=useNavigate();
  const dispatch=useDispatch();

  const handleLogOut=()=>{
    dispatch(startLogOut());
     navigate('/', {
                replace:true
              }) ;
  }
  const {name}=useSelector(state=>state.auth)
  const handleAddNew=()=>{
    dispatch(startNewNote());
  }
  return (
    <aside className='journal__sidebar animate__animated animate__backInLeft'>
        <div className='journal__sidebar-navbar'>
            <h3><i className='fa fa-moon '></i><span>{name}</span></h3>
            <button className='btn' onClick={handleLogOut}>Logout</button>
        </div>
        <div className='journal__new-entry' onClick={handleAddNew}>
          <i className='fa fa-calendar-plus fa-5x '></i>
          <p className='mt-5'>New Entry</p>
        </div>
        <JournalEntries />
    </aside>
  )
}
