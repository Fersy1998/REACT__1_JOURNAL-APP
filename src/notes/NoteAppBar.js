import React, { useEffect, useRef } from 'react'
import { useSelector } from 'react-redux';
import { useDispatch } from 'react-redux';

import { activeNote, startSaveNotes, startUploading } from '../actions/notes';
import { useForm } from '../hooks/useForm';

export const NoteAppBar = () => {
  const dispatch=useDispatch();
  const {active}=useSelector(state=>state.notes)
  
  const handleSave=()=>{
    dispatch(startSaveNotes({...active}))
  }
  const handlePictureClick=()=>{
      document.querySelector('#fileSelector').click();
  }
  const handleFileChange=(e)=>{
    const file=e.target.files[0];
    if(file){
      dispatch(startUploading(file));
    }
  
  }
  return (
    <div className='note__appbar'>
        <span>28 Agust 2020</span>
        <input type='file' name='file' id='fileSelector' style={{display:'none'}} onChange={handleFileChange}/>
        <div>
            <button className='btn' onClick={handlePictureClick}>
                Picture
            </button>
            <button className='btn' onClick={handleSave}>
                Save
            </button>
        
        </div>
       
    </div>
  )

}