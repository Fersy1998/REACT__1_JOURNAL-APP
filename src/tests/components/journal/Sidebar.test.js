import React from 'react';
import { Sidebar } from '../../../journal/Sidebar';
import '@testing-library/jest-dom';
import { mount } from "enzyme";

import { Provider } from "react-redux";
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { MemoryRouter } from "react-router-dom";
import {startNewNote, noteLogout } from '../../../actions/notes'

import {startLogOut } from '../../../actions/auth'
jest.mock("../../../actions/auth", ()=>({ 
    startLogOut : jest.fn() ,
    noteLogout : jest.fn() ,
}))
jest.mock("../../../actions/notes", ()=>({ 
    startNewNote : jest.fn() ,
}))
jest.setTimeout(9000);

const middlewares = [thunk];
const mockStore = configureStore(middlewares);


const initState={
    
        auth: {
          uid: 'tALkf4F0iidLjZAPc2suS9w2gzB3',
          name: 'Fersy'
        },
        ui: {
          loading: false,
          msgError: null
        },
        notes: {
          notes: [
            {
              id: 'wiE5xqElkfE0kJ1dA5kT',
              title: '',
              body: '',
              date: 1654747267844,
              url: 'https://res.cloudinary.com/dyyfc2gux/image/upload/v1654747279/original_-_2021-11-07T194242.528_yvsszx.jpg'
            }
          ],
          active: {
            id: 'wiE5xqElkfE0kJ1dA5kT',
            title: '',
            body: '',
            date: 1654747267844,
            url: 'https://res.cloudinary.com/dyyfc2gux/image/upload/v1654747279/original_-_2021-11-07T194242.528_yvsszx.jpg'
          }
        }
      
  }
  
let store=mockStore(initState);

store.dispatch=jest.fn()
  const wrapper=mount(
    <Provider store={store}>
        <MemoryRouter>
            <Sidebar />
        </MemoryRouter>
    </Provider> );


describe('Pruebas en el SideBar', ()=>{
    
    beforeEach(()=>{
        store=mockStore(initState);
        jest.clearAllMocks();
        
    })
    
    test('should mathc the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    
    })
    test('should startLogOut', () => { 
        wrapper.find('button').simulate('click');
        expect(startLogOut).toHaveBeenCalled();
    
    })
    test('should startNewNote', () => { 
        wrapper.find('.journal__new-entry').simulate('click');
        expect(startNewNote).toHaveBeenCalled();
    
    })
})