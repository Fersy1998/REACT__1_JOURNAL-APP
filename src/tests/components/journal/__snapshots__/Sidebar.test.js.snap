// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el SideBar should mathc the snapshot 1`] = `
<aside
  className="journal__sidebar animate__animated animate__backInLeft"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3>
      <i
        className="fa fa-moon "
      />
      <span>
        Fersy
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fa fa-calendar-plus fa-5x "
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries animate__animated animate__headShake"
  >
    <div
      className="journal__entry pointer"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://res.cloudinary.com/dyyfc2gux/image/upload/v1654747279/original_-_2021-11-07T194242.528_yvsszx.jpg)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        />
        <p
          className="journal__entry-content"
        />
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Wednesday
        </span>
        <p>
          8th
        </p>
      </div>
    </div>
  </div>
</aside>
`;
